mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp

```diff
@@ -0,0 +1,419 @@
+      return {};
+    selectedType = field;
+  }
+  return std::make_pair(selectedType, firstLevelIndex);
+}
+
+LogicalResult LLVM::GEPOp::ensureOnlyTypeSafeAccesses(
+    const MemorySlot &slot, SmallVectorImpl<MemorySlot> &mustBeSafelyUsed) {
+  if (getBase() != slot.ptr)
+    return success();
+  if (slot.elemType != getElemType())
+    return failure();
+  auto [reachedType, _] = computeReachedGEPType(*this);
+  if (!reachedType)
+    return failure();
+  mustBeSafelyUsed.emplace_back<MemorySlot>({getResult(), reachedType});
+  return success();
+}
+
+bool LLVM::GEPOp::canRewire(const DestructibleMemorySlot &slot,
+                            SmallPtrSetImpl<Attribute> &usedIndices,
+                            SmallVectorImpl<MemorySlot> &mustBeSafelyUsed) {
+  if (getBase() != slot.slot.ptr || slot.slot.elemType != getElemType())
+    return false;
+  auto [reachedType, firstLevelIndex] = computeReachedGEPType(*this);
+  if (!reachedType || !firstLevelIndex)
+    return false;
+  assert(slot.info.elementsPtrs.contains(firstLevelIndex));
+  if (!slot.info.elementsPtrs.at(firstLevelIndex).isa<LLVM::LLVMPointerType>())
+    return false;
+  mustBeSafelyUsed.emplace_back<MemorySlot>({getResult(), reachedType});
+  usedIndices.insert(firstLevelIndex);
+  return true;
+}
+
+DeletionKind LLVM::GEPOp::rewire(const DestructibleMemorySlot &slot,
+                                 DenseMap<Attribute, MemorySlot> &subslots) {
+  IntegerAttr firstLevelIndex = getIndices()[1].dyn_cast<IntegerAttr>();
+  const MemorySlot &newSlot = subslots.at(firstLevelIndex);
+
+  ArrayRef<int32_t> remainingIndices = getRawConstantIndices().slice(2);
+
+  // If the GEP would become trivial after this transformation, eliminate it.
+  if (llvm::all_of(remainingIndices,
+                   [](int32_t index) { return index == 0; })) {
+    getResult().replaceAllUsesWith(newSlot.ptr);
+    return DeletionKind::Delete;
+  }
+
+  // Rewire the indices by popping off the second index.
+  SmallVector<int32_t> newIndices;
+  newIndices.reserve(remainingIndices.size() + 1);
+  newIndices.push_back(0);
+  newIndices.append(remainingIndices.begin(), remainingIndices.end());
+  setRawConstantIndices(newIndices);
+
+  // Rewire the pointed type.
+  setElemType(newSlot.elemType);
+
+  // Rewire the pointer.
+  getBaseMutable().assign(newSlot.ptr);
+
+  return DeletionKind::Keep;
+}
+
+//===----------------------------------------------------------------------===//
+// Interfaces for destructible types
+//===----------------------------------------------------------------------===//
+
+Optional<DenseMap<Attribute, Type>> LLVM::LLVMStructType::destruct() {
+  int32_t index = 0;
+  Type i32 = IntegerType::get(getContext(), 32);
+  DenseMap<Attribute, Type> destructured;
+  for (Type elemType : getBody()) {
+    destructured.insert({IntegerAttr::get(i32, index), elemType});
+    index++;
+  }
+  return destructured;
+}
+
+Type LLVM::LLVMStructType::getTypeAtIndex(Attribute index) {
```

Comment:
No! It is actually different. While the map is only generated if the type is reasonably small (because if your array is a million element wide that would create a map with a million elements), getTypeAtIndex can get a type from a destructible type no matter how big it is.

Reply:
No no. `getTypeAtIndex` is for example used when you want to ensure the type safety of further uses. Say you have
`struct { i32, i32, array<10000000000 x i32> }`. It is definitely reasonable to SROA this at the first order. So you would ask for its destruction, leading to three new allocas: `i32`, `i32` and `array<10000000000 x i32>`. But you still need to check that those new allocas would be used in a type-safe manner. So if the last alloca had a further gep that would access the content of the array, you need to check that this access is type-safe. But of course it would be unreasonable to destruct array<10000000000 x i32>`. This is when getTypeAtIndex is useful: it will give you information about large aggregate types without having to generate the huge maps that destructing them would. This can be used even if the destruction request is not possible.

Reply:
Ok I understand. Are we checking the array accesses to make sure they are not out-of-bounds? Type safety on it's own should not be a concern as long as all accesses are inbounds? Or am I missing something?